src/circuit/gadgets/constraints.rs

```diff
@@ -430,6 +503,444 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
+        &mut cs.namespace(|| "Allocate num for diff_is_positive"),
+        &diff_is_positive,
+    )?;
+
+    let alloc_num_diff_is_not_negative = boolean_to_num(
+        &mut cs.namespace(|| "Allocate num for diff_is_not_negative"),
+        &diff_is_not_negative,
+    )?;
+
+    let mut comp_results = CompResults::default();
+    comp_results.add_clauses_comp(
+        Op2::Less.as_field(),
+        &alloc_num_diff_is_negative,
+        &g.true_num,
+        &g.false_num,
+    );
+    comp_results.add_clauses_comp(
+        Op2::LessEqual.as_field(),
+        &alloc_num_diff_is_not_positive,
+        &g.true_num,
+        &g.false_num,
+    );
+    comp_results.add_clauses_comp(
+        Op2::Greater.as_field(),
+        &alloc_num_diff_is_positive,
+        &g.false_num,
+        &g.true_num,
+    );
+    comp_results.add_clauses_comp(
+        Op2::GreaterEqual.as_field(),
+        &alloc_num_diff_is_not_negative,
+        &g.false_num,
+        &g.true_num,
+    );
+
+    let comparison_defaults = [&g.default_num, &g.default_num, &g.default_num];
+
+    let comp_clauses = [
+        &comp_results.same_sign[..],
+        &comp_results.a_neg_and_b_not_neg[..],
+        &comp_results.a_not_neg_and_b_neg[..],
+    ];
+
+    let comparison_result = multi_case_aux(
+        &mut cs.namespace(|| "comparison multicase results"),
+        op2,
+        &comp_clauses,
+        &comparison_defaults,
+        g,
+    )?;
+
+    let comp_val_same_sign_num = comparison_result.0[0].clone();
+    let comp_val_a_neg_and_b_not_neg_num = comparison_result.0[1].clone();
+    let comp_val_a_not_neg_and_b_neg_num = comparison_result.0[2].clone();
+    let is_comparison_tag = comparison_result.1.not();
+
+    let comp_val1 = pick(
+        &mut cs.namespace(|| "comp_val1"),
+        &a_negative_and_b_not_negative,
+        &comp_val_a_neg_and_b_not_neg_num,
+        &comp_val_a_not_neg_and_b_neg_num,
+    )?;
+    let comp_val2 = pick(
+        &mut cs.namespace(|| "comp_val2"),
+        &both_same_sign,
+        &comp_val_same_sign_num,
+        &comp_val1,
+    )?;
+
+    let comp_val_is_zero = alloc_is_zero(&mut cs.namespace(|| "comp_val_is_zero"), &comp_val2)?;
+
+    let comp_val = AllocatedPtr::pick(
+        &mut cs.namespace(|| "comp_val"),
+        &comp_val_is_zero,
+        &g.nil_ptr,
+        &g.t_ptr,
+    )?;
+
+    Ok((is_comparison_tag, comp_val, diff_is_negative.clone()))
+}
+
+// Enforce 0 <= num < 2ˆn.
+pub fn enforce_n_bits<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    num: AllocatedNum<F>,
+    n: usize,
+) -> Result<(), SynthesisError> {
+    let num_bits = num.to_bits_le(&mut cs.namespace(|| "u64 remainder bit decomp"))?;
+    let v = num_bits[(n + 1)..255].to_vec();
+    boolean_summation(&mut cs.namespace(|| "add all MSBs"), &v, &g.false_num);
+    Ok(())
+}
+
+// Enforce div and mod operation for U64. We need to show that
+// arg1 = q.arg2 + r, such that 0 <= r < b.
```

Comment:
This might be clearer if you wrote `q * arg2`.

Reply:
It's also unclear what `b` represents.  I can guess, but we should probably write it out here.